### DISCUSSION TITLE
Replace a bunch of keywords at once?

### DISCUSSION THREAD

**Discussion by aryez91** [dnGrep/dnGrep#761] [Q&A] [answered]
Hi people,

Is there any way to replace a bunch of different words at one search-and-replace operation?

Say I have two lists of words: one for search criteria, and another one for corresponding substitutions... Like so:

**List:** Example1, Example2, Example3....
**Substitution list**: Example1Substitution, Example2Substitutions, Example3Substitutions.... etc...

The long way is to search one by one:
Search for: *Example1*
Replace with:  *Example1Substitute*

Search for: *Example2*
Replace with *Example2Substitute*

etc...

Can it be done with one go?

Search for: (list)
Replace with: (substitution list) 

Help appreciated.

**Comment by fade2gray** [comment, marked as answer]
Not as it stands, but Doug is thinking about adding a scripting feature in a future release https://github.com/dnGrep/dnGrep/discussions/755#discussioncomment-3562677

**Comment by jcmil101** [comment]
I too would like to be able to do a SAR with delimited literal text in non-overlapping search via lists or .csv etc.
Scripting would be a great addition.

**Comment by vatterspun** [comment]
This can be done in batch via the command line using a neighboring program on the Windows side called FNR (Find and Replace).  I've currently got something to highlight words in a document (markdown highlighting using two equals signs):

`"C:\temp\FNR\fnr.exe" --cl --dir "C:\temp\Automation\HTML-Input" --fileMask "*.md" --excludeFileMask "*.dll, *.exe" --includeSubDirectories --caseSensitive --find "appscan" --replace "==appscan=="`

I generate these using a simple spreadsheet that puts the find and replace into separate columns.  I can post that if anyone's interested.  Then you save it as a batch file and run it.  Works crazy fast.  

Note that it doesn't work if you need to replace a " character.  I've been meaning to post a bug report about that.

**Reply by jcmil101** [reply]
Sounds good, I'd like to see your file. :)
Is there a way to do a non-overlapping SAR with this method or will a non-match halt the processing?

**Reply by vatterspun** [reply]
>> I generate these using a simple spreadsheet that puts the find and replace into separate columns.
>
> I'd like to see your file :)

Sure thing ... the spreadsheet is linked below. It's in LibreOffice format but should open fine in whatever spreadsheet program you've got:

[replace syntax.ods](https://github.com/dnGrep/dnGrep/files/10210188/replace.syntax.ods)

... and please do let me know if or how you use it.  I feel like I'm on to something with this but lack the time to really dig into it.

---

As far as the non-overlapping SAR, I really don't know.  That's a bit above my expertise.

**Reply by jcmil101** [reply]
Thanks for posting that.  I will have to see if I can base a non-overlapping SAR on this formula.  If I can manage it I'll post back :)

**Comment by fade2gray** [comment]
Here's a simple AutoHotkey script. You can either use single search and replace terms or use lists of comma separated terms.

It will create a copy of the original file appended with `.SaR`, i.e. `TestFile.txt.SaR`

![Search and Replace](https://user-images.githubusercontent.com/1332009/203445681-0efd2951-e001-4d2b-b281-2098c4f329b7.png)

<details>
<summary>Click to reveal script</summary>
  
```ahk
#NoEnv
#SingleInstance, Force
SendMode, Input
SetBatchLines, -1
SetWorkingDir, %A_ScriptDir%

Gui, Add, Text,,% "Enter full path to file (do not enclose in quotes):"
Gui, Add, Edit, w250 vFile
Gui, Add, Text,,% "Enter comma separated list of search terms:"
Gui, Add, Edit, w250 vSearch
Gui, Add, Text,,% "Enter comma separated list of replacement terms:"
Gui, Add, Edit, w250 vReplace
Gui, Add, Button, gSearchAndReplace,% "Search && Replace"
Gui, Add, Button, xp+110 yp,% "Exit"
Gui, Show,,% "Search & Replace"
Return

SearchAndReplace:
  Gui, Submit, NoHide
  Search := StrSplit(Search, ",")
  Replace := StrSplit(Replace, ",")
  If (Search.Count() != Replace.Count())
    MsgBox,% "Different number of search and replace terms."
  FileRead, content,% file
  Loop,% Search.count()
    content := StrReplace(content, Search[A_Index], Replace[A_Index])
  FileAppend,% content,% file ".SaR"
  MsgBox,% "Finished."
Return

ButtonExit:
GuiEscape:
GuiClose:
ExitApp
```
</details>

Attached is a compiled version of the script if you don't have AutoHotkey installed.
[Search and Replace.exe.zip](https://github.com/dnGrep/dnGrep/files/10071966/Search.and.Replace.exe.zip)

**Reply by jcmil101** [reply]
This is useful.  What do you recommend for non-overlapping terms & through multiple files?

**Reply by fade2gray** [reply]
I'm by no means an expert in using AHK, but there's a wealth of excellent help and advice to be found on the [AutoHotkey Community](https://www.autohotkey.com/boards/viewforum.php?f=4).

**Reply by jcmil101** [reply]
Coincidentally I've just started looking at that.  :) 
I'm wondering what the best way to add sth like ELSE...IF is.

**Reply by fade2gray** [reply]
It's all in the documentation ;) https://www.autohotkey.com/docs/commands/IfExpression.htm

**Reply by jcmil101** [reply]
That does look like it might do what I'd like it to.  :)

**Comment by doug24** [comment]
[v3.1.197](https://github.com/dnGrep/dnGrep/releases/tag/v3.1.197.0) has been released with scripting.  Using scripts, you can do multiple searches and replaces with different search patterns and replace text.  Each search and replace is a separate operation, mainly because of how dnGrep modifies the files during a replace, but also to avoid any problems with overlapping search results.  A dnGrep script may look like:

```
set folder C:\Repos\testFiles\test\scriptTest\
set pathtomatch *.cs
set searchfor string
set replacewith String
search
replace
set searchfor bool
set replacewith Boolean
search
replace
```

**Reply by jcmil101** [reply]
Great news!  Thanks!  Sounds like I need to start playing with this right away :)
